utils/spdx/src/main/kotlin/SpdxExpression.kt

```diff
@@ -314,7 +314,18 @@ class SpdxCompoundExpression(
 
     override fun validChoicesForDnf(): Set<SpdxExpression> =
         when (operator) {
-            SpdxOperator.AND -> setOf(decompose().reduce(SpdxExpression::and))
+            SpdxOperator.AND -> {
+                val leftChoices = left.validChoicesForDnf()
+                val rightChoices = right.validChoicesForDnf()
+
+                // Cartesian product of choices on the left and right.
+                leftChoices.flatMapTo(mutableSetOf()) { leftChoice ->
+                    rightChoices.map { rightChoice ->
+                        leftChoice and rightChoice
+                    }
+                }
+            }
```

Comment:
> Therefore I believe the performance should be examined to ensure the change does not make that worse.

I agree, but since I've so far not experienced any real-life performance issues with this code, I'm not in a good position to examine the performance, and would appreciate some assistance here.

Anyway, to me correctness is more important than performance here, and this fixes a rather severe (IMO) bug with valid choice for `AND` expressions. So I'd prefer to merge this rather sooner than later, and fix any performance issues in the aftermath.

Reply:
> Anyway, to me correctness is more important than performance here,

I've been talking about a situation were execution takes "forever". Was that clear to you? (just to double check)

Reply:
> I'll take a look if I can find the reproducer.

Thanks.

> I've been talking about a situation were execution takes "forever". Was that clear to you? (just to double check)

Yes. But I'm considering that to be an edge-case (though with a severe impact), whereas the fix for choices with `AND` expressions is probably needed in way more cases.

Reply:
I'd prefer checking performance impact first, but I'm ok if you guys decide to merge it.

Reply:
In lack of a better performance check, or at least input data to operate on, all I can say is that I've looked at the run times of `SpdxExpressionTest` which did not seem to have changed with any significance.

Reply:
IIRC the performance issue happens only if the code path covers this line: https://github.com/oss-review-toolkit/ort/blob/920fd0ccb5667526003ec17fe40d9117e83e990f/model/src/main/kotlin/licenses/ResolvedLicenseInfo.kt#L84